### ISSUE TITLE
Feature: Multi Node / Multi Cluster Support

### ISSUE THREAD

**Issue by smerschjohann** [rancher-sandbox/rancher-desktop#588]
Hi,

I'm a huge fan of k3d as it allows to create multiple clusters as well as multiple nodes in that cluster.
It would be nice if something similar would be possible using the WSL2 and rancher-desktop as well.

I currently use the k3d with the following parameters:

```
docker create volume mycluster-data
k3d cluster create mycluster \
        -a 2 \
        -i docker.io/rancher/k3s:v1.19.9-k3s1 \
        -p 80:80@loadbalancer -p443:443@loadbalancer \
        --registry-use k3d-registry \
        -v mycluster-data:/data
```

This means that for easier usage, all nodes share the /data volume. So to speak a "NFS" and are able to use a custom registry.

So it would be nice to have this options in rancher-desktop as well.

**Comment by smerschjohann** [comment]
I scanned through the current codebase and I see that it is neither prepared for multiple clusters nor multiple nodes in the current state.

It could be resolved by changing it in that way, that it does not use a single WSL2 distribution (let's name them wsl2 instances), but instead create them by some convention like:

- rancher-desktop-data
- rancher-desktop-CLUSTERNAME1-0
- rancher-desktop-CLUSTERNAME1-1
- rancher-desktop-CLUSTERNAME2-0

The data wsl2 instance is only used to mount "volumes" into the clusters which can be used as shared storage like NFS etc. At least in my scenarios it makes development a lot easier.

What do you think?

**Comment by jandubois** [comment]
Hi @smerschjohann!

Thank you for your feedback! I agree that it will be nice to support multi-node clusters (and personally I would like to run mixed-architecture clusters, where one node could run ARM containers using emulation), but there is still a lot of functionality that needs to be added for single-cluster usage, and that is what we'll have to focus on in the near future.

Getting requests for multi-node clusters certainly helps us to see that there is demand for this, and will be used to prioritize future work.

**Comment by smerschjohann** [comment]
yes, I understand that. But I'm not sure if this functionality must be kept in mind, when other features are implemented. Otherwise it might be a lot more difficult to integrate this feature later on.

**Comment by jandubois** [comment]
We do keep this in mind. For example we want to eventually support a multi-architecture cluster, where you can run both amd64 and arm64 container images. That would require at least 2 nodes, one for each architecture.

Could you provide some more information about your use cases? It is currently not clear to me if we would implement multiple nodes by running them inside separate WSL distros, or if we could run additional nodes in containers inside the same distro that runs the control plane. Would this make any difference for your use case?

I've been re-reading this whole issue and noticed that you were also asking about multi-cluster support. Can you expand on that as well? Do you want to work at one cluster at a time, or do you expect to have multiple clusters (with multiple nodes each) running on your desktop, and if yes, how would they need to be connected?

I don't really see multi-cluster support in the short-to-medium future though, given how many other features are still outstanding for single-node (and then multi-node) clusters. But, as always, priorities are driven by user feedback.

**Comment by smerschjohann** [comment]
I like the idea of supporting multiple architectures, especially since multiple cloud providers are investing on that front. I'm currently playing around with the OCI Ampere A1, I think it will be become more prominent in the near future.

My idea was the use of multiple WSL distros, each running a single k3s binary. They may share some data paths using the /mnt/wsl directory. The same endresult might be accomplished in a single distro, but I guess this will be more difficult to implement. The other solution will profit directly from the abstractions the WSL can give us.

Yes, I would like multiple cluster support, with k3d it helps me in that way, that I can have completly different projects/customers  each having their own cluster. Each project can have multiple namespaces, ingress routers, NetworkPolicies, ... and I can switch between them by turning off the one cluster and starting the other one.

So my use case is more like a "project and customer switch" than testing "Multi-Cluster" / "Multi-Cloud" solutions, which might be cool in some later projects, but is not in my focus right now. 

Another option for my "switch projects" cluster use case: I just need one cluster at a time, so if the "data distro" or the path in that distro can be switched, it could be enough as that will result in the switch of the underlying sqlite db (kubernetes state), PVs and other states of that cluster. Nonetheless this solution is very dirty and I think not favorable.

**Comment by jandubois** [comment]
> So my use case is more like a "project and customer switch" than testing "Multi-Cluster" / "Multi-Cloud" solutions, which might be cool in some later projects, but is not in my focus right now.

Thanks, that makes a lot of sense. I also worry that most laptop/desktop computers may not have enough power to really run a multi-cluster-multi-node setup, so it might be of limited use. When you run multiple clusters simultaneously, it may be time to switch to a cloud provider...

**Comment by smerschjohann** [comment]
> When you run multiple clusters simultaneously, it may be time to switch to a cloud provider...

I agree. It might help in some edge cases, but in most cases, one can test/develop the same things in a single cluster. In all other cases, it should be possible to switch to other solutions. - no need to complicate the UI

**Comment by Issif** [comment]
Hello, here my use-cases which need multi-nodes:
- ensure affinity works for avoiding to run several pods on same nodes
- run tests with replicated databases accross nodes (I'm trying to do so with rqlite on a k8s)

**Comment by jandubois** [comment]
@Issif That you for showing your use-cases. For the near-future, you will need to use `k3d` to run these.

We are aiming to be able to run `k3d` with just Rancher Desktop in the next release (probably with some limitations regarding volume mounts).

**Comment by ankurpshah** [comment]
+1

**Comment by YourTechBud** [comment]
I noticed rancher desktop has added support for k3d. Can we expect multiple clusters via k3d soon?

**Comment by jandubois** [comment]
@YourTechBud You should be able to create multiple cluster via `k3d` right now (you must use the `moby` runtime for this). If something isn't working correctly, please file an issue!

**Comment by shinebayar-g** [comment]
Hmm. Multi node, multi cluster features would be such a great addition. I prefer Rancher Desktop UI over Docker Desktop but both solutions only support single cluster with single node. I guess I'm going back to kind/minikube 🥹
